File: kotlin/src/main/kotlin/com/bhengubv/circleai/hosting/Neuron.kt
// Neuron.kt
//
// The Neuron — Kotlin port of CircleAI.Hosting.Neuron. The Kotlin hosting
// AIService is a thin port (it owns an injected generatorFactory, not a
// selector/loader), so the two-slot residency is adapted to that idiom: the
// concierge routes a turn, an injected selector resolves the capability to a
// model id, and an injected specialistFactory builds the specialist generator.
//
// Contents: the concierge router + gate, the RAM admission gate
// (ResidentSlotManager), and the host-neutral NeuronNode facade.

package com.bhengubv.circleai.hosting

import com.bhengubv.circleai.inference.IChatGenerator
import com.bhengubv.circleai.models.ChatMessage
import com.bhengubv.circleai.selector.ChatCapability
import com.bhengubv.circleai.selector.ModelSelection
import kotlinx.coroutines.flow.Flow
import java.util.UUID

// =====================================================================
// Concierge router + gate
// =====================================================================

/** Which resident organ should serve a turn. Mirrors the Organ enum. */
enum class Organ { GENERALIST, SPECIALIST }

/** What the router sees for one turn. Mirrors RouteContext. */
data class RouteContext(
    val query: String,
    val hasImage: Boolean = false,
    val promptChars: Int? = null,
)

/** The router's per-turn verdict. Mirrors RouteDecision. */
data class RouteDecision(
    val organ: Organ,
    /** ChatCapability flags the specialist must satisfy (ignored for the generalist). */
    val capability: Int,
    val reason: String,
)

/** Per-turn router contract. Mirrors INeuronRouter. */
fun interface INeuronRouter {
    fun route(context: RouteContext): RouteDecision
}

/**
 * Safety/capability veto over a specialist pick. Mirrors NeuronGate: when the
 * predicate rejects a decision the router demotes it to the generalist, so a
 * turn is never blocked — the floor always answers.
 */
class NeuronGate(
    private val allowSpecialist: (RouteDecision) -> Boolean = { true },
) {
    fun allows(decision: RouteDecision): Boolean = allowSpecialist(decision)
}

/**
 * The default concierge router. Cheap heuristics, in priority order:
 *   image bytes     → Specialist(VISION)
 *   long prompt     → Specialist(LONG_CONTEXT)   (≥ longContextChars)
 *   reasoning cues  → Specialist(REASONING)
 *   otherwise       → Generalist(DEFAULT)
 * A NeuronGate veto demotes any specialist pick back to the generalist.
 * Mirrors HeuristicNeuronRouter.
 */
class HeuristicNeuronRouter(
    private val longContextChars: Int = 4000,
    private val gate: NeuronGate = NeuronGate(),
) : INeuronRouter {
    override fun route(context: RouteContext): RouteDecision {
        val d = classify(context)
        return if (d.organ == Organ.SPECIALIST && !gate.allows(d)) {
            RouteDecision(Organ.GENERALIST, ChatCapability.DEFAULT, "gate-vetoed → generalist")
        } else {
            d
        }
    }

    private fun classify(context: RouteContext): RouteDecision {
        if (context.hasImage) {
            return RouteDecision(Organ.SPECIALIST, ChatCapability.VISION, "image present")
        }
        val chars = context.promptChars ?: context.query.length
        if (chars >= longContextChars) {
            return RouteDecision(Organ.SPECIALIST, ChatCapability.LONG_CONTEXT, "long prompt ($chars chars)")
        }
        val q = context.query.lowercase()
        if (REASONING_CUES.any { q.contains(it) }) {
            return RouteDecision(Organ.SPECIALIST, ChatCapability.REASONING, "reasoning cue")
        }
        return RouteDecision(Organ.GENERALIST, ChatCapability.DEFAULT, "default")
    }

    private companion object {
        val REASONING_CUES = listOf(
            "debug", "stack trace", "solve", "prove", "reason", "analy",
            "calculate", "equation", "step by step", "algorithm",
            "why does", "derive", "diagnose",
        )
    }
}

// =====================================================================
// RAM admission gate — the specialist slot beside the generalist floor
// =====================================================================

/** Outcome of an admission attempt. Mirrors SlotOutcome. */
enum class SlotOutcome { ADMITTED, ALREADY_RESIDENT, INSUFFICIENT_RAM, BUILD_FAILED }

/** Result of ensureSpecialist. Mirrors SlotAdmission. */
data class SlotAdmission(val outcome: SlotOutcome, val generator: IChatGenerator?)

/**
 * Owns the single hot-swappable specialist slot beside the always-warm
 * generalist floor. Admission = generalistReserved + estimatedBytes ≤ ceiling.
 * A different pick evicts the incumbent first (one specialist at a time), and
 * the specialist is evicted first under memory pressure so the generalist never
 * drops. Mirrors ResidentSlotManager (modeled on the server's
 * ModelLifecycleManager).
 */
class ResidentSlotManager(
    private val generalistReservedBytes: Long,
    private val ramAvailableBytes: () -> Long,
) {
    private var specialist: IChatGenerator? = null
    private var specialistModelId: String? = null

    /** The resident specialist's model id, or null if the slot is empty. */
    val residentSpecialistModelId: String? get() = specialistModelId

    /** The resident specialist generator, or null. */
    val residentSpecialist: IChatGenerator? get() = specialist

    /**
     * Ensure [selection] is the resident specialist, building it via [build] if
     * needed. Admission-gated on RAM; a different pick evicts the incumbent.
     * Never throws on denial — returns the outcome so the caller can fall back
     * to the generalist floor.
     */
    fun ensureSpecialist(
        selection: ModelSelection,
        build: (String) -> IChatGenerator?,
    ): SlotAdmission {
        val id = selection.modelId
        val current = specialistModelId
        if (current != null && current.equals(id, ignoreCase = true)) {
            return SlotAdmission(SlotOutcome.ALREADY_RESIDENT, specialist)
        }

        // RAM admission gate: reserve the floor, then check the specialist fits.
        val needed = generalistReservedBytes + maxOf(0L, selection.estimatedBytes)
        if (needed > ramAvailableBytes()) {
            return SlotAdmission(SlotOutcome.INSUFFICIENT_RAM, null)
        }

        // Evict the incumbent (one specialist at a time) before building the new.
        evictSpecialist()

        val built = build(id) ?: return SlotAdmission(SlotOutcome.BUILD_FAILED, null)
        specialist = built
        specialistModelId = id
        return SlotAdmission(SlotOutcome.ADMITTED, built)
    }

    /** Drop the resident specialist (the generalist floor is untouched). */
    fun evictSpecialist() {
        val g = specialist
        specialist = null
        specialistModelId = null
        if (g is AutoCloseable) runCatching { g.close() }
    }
}

// =====================================================================
// NeuronNode facade
// =====================================================================

/**
 * A complete, host-neutral Neuron node over an [IAIService] brain. Composes the
 * concierge (a router-gated two-slot AIService) behind the host-neutral
 * [IChatRuntime] / [IPersistableChatRuntime] seam, and exposes the underlying
 * brain so a CompanionSession can still sit on top. Persists the generalist
 * floor's session for OOM/restart survival (the specialist is rebuildable from
 * the registry). Mirrors NeuronNode : IChatRuntime, IPersistableChatRuntime.
 */
class NeuronNode(
    private val brain: IAIService,
    sessionSnapshotPath: String? = null,
) : IPersistableChatRuntime {

    private val snapshotPathValue: String? =
        sessionSnapshotPath
            ?: System.getProperty("java.io.tmpdir")?.let { dir ->
                (if (dir.endsWith("/") || dir.endsWith("\\")) dir else "$dir/") +
                    "circleai-neuron-session.bin"
            }

    /** The underlying brain, so CompanionSession can compose over the Neuron. */
    val brainService: IAIService get() = brain

    override val id: String get() = "circleai-neuron"

    override val engineLabel: String
        get() {
            val mid = (brain as? AIService)?.resolvedModelIdValue
            return if (!mid.isNullOrEmpty()) "circleai-neuron:$mid" else "circleai-neuron"
        }

    override val isReady: Boolean get() = brain.isReady

    override val statusMessage: String get() = if (brain.isReady) "ready" else "loading model…"

    override val sessionSnapshotPath: String? get() = snapshotPathValue

    override fun streamAsync(turns: List<ChatTurn>): Flow<String> {
        val messages = turns.map { t ->
            ChatMessage(id = UUID.randomUUID().toString(), role = t.role, content = t.content)
        }
        return brain.streamAsync(messages)
    }

    override suspend fun saveSessionAsync(path: String): Boolean = brain.saveSessionAsync(path)

    override suspend fun loadSessionAsync(path: String): Boolean = brain.loadSessionAsync(path)
}

File: rust/src/hosting/neuron.rs
//! neuron.rs
//!
//! The Neuron — concierge router + two-slot residency + host-neutral facade.
//! Port of `CircleAI.Hosting.Neuron`, adapted to the Rust sync hosting layer.
//!
//! The C# two-slot reuses `IModelSelector.BestFit` + `IModelLoader`; Rust's
//! sync hosting layer erases the generator behind `IHostChatGenerator` and its
//! selector API diverges, so the specialist path is expressed as two injected
//! closures: a `neuron_selector` (capability -> [`SpecialistPick`], the BestFit
//! analog) and a `specialist_builder` (model id -> generator, the loader analog).
//! The specialist slot holds an `Arc` so it can be cloned out of the lock.

use std::sync::{Arc, Mutex};

use crate::inference::ChatMessage;
use crate::selector::ChatCapability;

use super::chat_runtime::{ChatTurn, IChatRuntime, IPersistableChatRuntime};
use super::service::{HostingError, IAIService, IHostChatGenerator};

// ─────────────────────────────────────────────────────────────────────────────
// Concierge router
// ─────────────────────────────────────────────────────────────────────────────

/// Which organ answers a turn.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum Organ {
    /// The always-warm generalist (the floor — never evicted).
    Generalist,
    /// A capability-matched specialist in the second slot.
    Specialist,
}

/// Inputs the concierge classifies for a single turn.
#[derive(Debug, Clone)]
pub struct RouteContext {
    pub query: String,
    pub has_image: bool,
}

/// The concierge's per-turn decision.
#[derive(Debug, Clone)]
pub struct RouteDecision {
    pub organ: Organ,
    pub capability: ChatCapability,
    pub reason: String,
}

impl RouteDecision {
    /// Route to the always-warm generalist.
    pub fn generalist(reason: impl Into<String>) -> Self {
        Self { organ: Organ::Generalist, capability: ChatCapability::DEFAULT, reason: reason.into() }
    }
    /// Route to a capability-matched specialist.
    pub fn specialist(capability: ChatCapability, reason: impl Into<String>) -> Self {
        Self { organ: Organ::Specialist, capability, reason: reason.into() }
    }
}

/// The concierge's decision layer. Mirrors `INeuronRouter`.
pub trait INeuronRouter: Send + Sync {
    fn route(&self, ctx: &RouteContext) -> RouteDecision;
}

/// Guardrail checkpoint. A `None` predicate applies no veto — the honest default.
pub struct NeuronGate {
    allow_specialist: Option<Box<dyn Fn(&str) -> bool + Send + Sync>>,
}

impl NeuronGate {
    /// Builds a gate. `None` applies no veto.
    pub fn new(allow_specialist: Option<Box<dyn Fn(&str) -> bool + Send + Sync>>) -> Self {
        Self { allow_specialist }
    }
    /// A gate that never vetoes.
    pub fn open() -> Self {
        Self { allow_specialist: None }
    }
    /// Apply the guardrail, returning the effective decision.
    pub fn apply(&self, decision: RouteDecision, ctx: &RouteContext) -> RouteDecision {
        if decision.organ == Organ::Specialist {
            if let Some(pred) = &self.allow_specialist {
                if !pred(&ctx.query) {
                    return RouteDecision::generalist("gate: specialist vetoed -> generalist");
                }
            }
        }
        decision
    }
}

const REASONING_CUES: &[&str] = &[
    "prove", "solve", "calculate", "derive", "algorithm", "complexity", "debug",
    "stack trace", "refactor", "regex", "step by step", "step-by-step", "theorem",
    "equation", "big-o", "big o",
];

/// Default router: modality (image -> vision), length (long -> long-context), and
/// reasoning cues (-> reasoning); else the generalist. Mirrors
/// `HeuristicNeuronRouter`.
pub struct HeuristicNeuronRouter {
    gate: NeuronGate,
    long_context_chars: usize,
}

impl HeuristicNeuronRouter {
    /// Builds the default router. `long_context_chars == 0` uses 4000.
    pub fn new(gate: NeuronGate, long_context_chars: usize) -> Self {
        Self {
            gate,
            long_context_chars: if long_context_chars == 0 { 4000 } else { long_context_chars },
        }
    }

    fn classify(&self, ctx: &RouteContext) -> RouteDecision {
        if ctx.has_image {
            return RouteDecision::specialist(ChatCapability::VISION, "image attached -> vision specialist");
        }
        if ctx.query.chars().count() >= self.long_context_chars {
            return RouteDecision::specialist(ChatCapability::LONG_CTX, "long prompt -> long-context specialist");
        }
        let lower = ctx.query.to_lowercase();
        for cue in REASONING_CUES {
            if lower.contains(cue) {
                return RouteDecision::specialist(ChatCapability::REASONING, "reasoning cue -> reasoning specialist");
            }
        }
        RouteDecision::generalist("no specialist cue -> generalist")
    }
}

impl INeuronRouter for HeuristicNeuronRouter {
    fn route(&self, ctx: &RouteContext) -> RouteDecision {
        self.gate.apply(self.classify(ctx), ctx)
    }
}

// ─────────────────────────────────────────────────────────────────────────────
// Two-slot residency
// ─────────────────────────────────────────────────────────────────────────────

/// The BestFit analog: a capability resolved to a concrete specialist model +
/// its estimated footprint (for the RAM gate).
#[derive(Debug, Clone)]
pub struct SpecialistPick {
    pub model_id: String,
    pub estimated_bytes: i64,
}

/// Outcome of a specialist-slot admission attempt.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum SlotOutcome {
    Admitted,
    AlreadyResident,
    InsufficientRam,
    BuildFailed,
}

/// Result of [`ResidentSlotManager::ensure_specialist`].
pub struct SlotAdmission {
    pub outcome: SlotOutcome,
    pub generator: Option<Arc<dyn IHostChatGenerator>>,
    pub message: String,
}

/// Manages one evictable specialist slot beside the always-warm generalist floor
/// (held by `AIService`). Only the generalist's reserved footprint counts against
/// the RAM gate. Mirrors `ResidentSlotManager`.
pub struct ResidentSlotManager {
    generalist_reserved_bytes: i64,
    ram_available: Box<dyn Fn() -> i64 + Send + Sync>,
    slot: Mutex<Option<(String, Arc<dyn IHostChatGenerator>)>>,
}

impl ResidentSlotManager {
    /// Builds a manager. `ram_available` supplies the live RAM ceiling.
    pub fn new(generalist_reserved_bytes: i64, ram_available: Box<dyn Fn() -> i64 + Send + Sync>) -> Self {
        Self {
            generalist_reserved_bytes: generalist_reserved_bytes.max(0),
            ram_available,
            slot: Mutex::new(None),
        }
    }

    /// The resident specialist's model id, or `None`.
    pub fn resident_specialist_model_id(&self) -> Option<String> {
        self.slot.lock().unwrap().as_ref().map(|(id, _)| id.clone())
    }

    /// Ensure a specialist for `pick` is resident, building it via `build` when
    /// needed. Admission gate: the generalist floor + the specialist footprint
    /// must fit under the RAM ceiling. On denial / build failure the slot is left
    /// empty and the caller answers from the generalist.
    pub fn ensure_specialist(
        &self,
        pick: &SpecialistPick,
        build: &dyn Fn(&str) -> Option<Arc<dyn IHostChatGenerator>>,
    ) -> SlotAdmission {
        let mut slot = self.slot.lock().unwrap();

        if let Some((id, gen)) = slot.as_ref() {
            if id.eq_ignore_ascii_case(&pick.model_id) {
                return SlotAdmission {
                    outcome: SlotOutcome::AlreadyResident,
                    generator: Some(gen.clone()),
                    message: format!("Specialist '{}' already resident.", pick.model_id),
                };
            }
        }

        let ceiling = (self.ram_available)().max(0);
        let needed = self.generalist_reserved_bytes + pick.estimated_bytes.max(0);
        if ceiling > 0 && needed > ceiling {
            return SlotAdmission {
                outcome: SlotOutcome::InsufficientRam,
                generator: None,
                message: format!(
                    "Specialist '{}' needs {} MiB; device ceiling {} MiB.",
                    pick.model_id,
                    needed >> 20,
                    ceiling >> 20
                ),
            };
        }

        // Only one specialist slot — evict the incumbent before building.
        *slot = None;

        match build(&pick.model_id) {
            Some(gen) => {
                *slot = Some((pick.model_id.clone(), gen.clone()));
                SlotAdmission {
                    outcome: SlotOutcome::Admitted,
                    generator: Some(gen),
                    message: format!("Specialist '{}' resident.", pick.model_id),
                }
            }
            None => SlotAdmission {
                outcome: SlotOutcome::BuildFailed,
                generator: None,
                message: format!("Specialist '{}' build failed.", pick.model_id),
            },
        }
    }

    /// Evict the specialist (the generalist floor is never touched).
    pub fn evict_specialist(&self) {
        *self.slot.lock().unwrap() = None;
    }
}

// ─────────────────────────────────────────────────────────────────────────────
// NeuronNode facade
// ─────────────────────────────────────────────────────────────────────────────

/// Host-neutral `IChatRuntime` + `IPersistableChatRuntime` over an `IAIService`
/// brain. Streaming rides the brain's full pipeline (enrichment + concierge
/// routing + two-slot residency). Mirrors `NeuronNode`.
pub struct NeuronNode {
    brain: Arc<dyn IAIService>,
    id: String,
    snapshot_path: String,
}

impl NeuronNode {
    /// Builds the facade. `id` `None`/empty defaults to `"circleai-neuron"`.
    pub fn new(brain: Arc<dyn IAIService>, id: Option<String>, snapshot_path: Option<String>) -> Self {
        Self {
            brain,
            id: id
                .filter(|s| !s.trim().is_empty())
                .unwrap_or_else(|| "circleai-neuron".to_string()),
            snapshot_path: snapshot_path.unwrap_or_else(default_snapshot_path),
        }
    }

    /// The on-device brain. A companion session consumes it unchanged.
    pub fn brain(&self) -> &Arc<dyn IAIService> {
        &self.brain
    }
}

impl IChatRuntime for NeuronNode {
    fn id(&self) -> String {
        self.id.clone()
    }
    fn engine_label(&self) -> String {
        "CircleAI Neuron".to_string()
    }
    fn is_ready(&self) -> bool {
        self.brain.is_ready()
    }
    fn status_message(&self) -> String {
        if self.brain.is_ready() {
            "ready".to_string()
        } else {
            "loading model…".to_string()
        }
    }
    fn stream(&self, messages: &[ChatTurn]) -> Result<Vec<String>, HostingError> {
        let mapped: Vec<ChatMessage> = messages
            .iter()
            .map(|t| ChatMessage::new(t.role.clone(), t.content.clone()))
            .collect();
        self.brain.stream(&mapped, None)
    }
}

impl IPersistableChatRuntime for NeuronNode {
    fn session_snapshot_path(&self) -> Option<String> {
        Some(self.snapshot_path.clone())
    }
    // The sync hosting generator (`IHostChatGenerator`) is erased and carries no
    // KV snapshot, so session persistence is a documented no-op in this port —
    // the API seam is present for parity.
    fn save_session(&self, _path: &str) -> Result<bool, HostingError> {
        Ok(false)
    }
    fn load_session(&self, _path: &str) -> Result<bool, HostingError> {
        Ok(false)
    }
}

fn default_snapshot_path() -> String {
    let base = std::env::var("LOCALAPPDATA")
        .ok()
        .filter(|s| !s.is_empty())
        .unwrap_or_else(|| std::env::temp_dir().to_string_lossy().to_string());
    format!("{base}/CircleAI/sessions/active.session")
}
